### ISSUE TITLE
Mathlib `have` tactic support

### ISSUE THREAD

**Issue by hrmacbeth** [Paper-Proof/paperproof#38]
The Mathlib `have` tactic does not seem to be supported yet, is that right?

The following two code snippets are mathematically equivalent, as far as I'm concerned, and it would be nice if they produced the same display:
```lean
import Mathlib.Tactic.Positivity
import Mathlib.Tactic.Linarith
import Mathlib.Tactic.Have
import Paperproof

example {x y : ℚ} : 2 * x * y ≤ x ^ 2 + y ^ 2 := by
  have H : 0 ≤ (x - y) ^ 2 := by
    positivity
  linarith

example {x y : ℚ} : 2 * x * y ≤ x ^ 2 + y ^ 2 := by
  have H : 0 ≤ (x - y) ^ 2 
  · positivity
  linarith
```
But they generate different Paperproof output -- I think the one above (for core `have`) is much better than the one below (for Mathlib `have`).
<img width="328" alt="Screenshot 2024-03-06 at 11 05 11 PM" src="https://github.com/Paper-Proof/paperproof/assets/25316162/a68bf9f5-03f6-4c2e-acf3-4b60c0e69a5b">
<img width="406" alt="Screenshot 2024-03-06 at 11 04 49 PM" src="https://github.com/Paper-Proof/paperproof/assets/25316162/8aef1ca6-bb76-47d8-933c-af0f4e4c56b1">
I'm not sure if there is a principled reason for the difference or if it's just that Mathlib `have` is not yet explicitly supported.

**Comment by antonkov** [comment]
Hi Heather, that's indeed an interesting example.

**tldr;**
Paperproof doesn't explicitly check tactic names and tries to be as generic as possible.
However the heuristic we apply to prettify `have`-like tactics isn't generic enough yet to
capture the second case. I agree that the Paperproof output for the two proofs should be
the same and _will think how we can generalise the heuristic further_.

**details**
The reason regular `have` produces the better output is because we have the following heuristic: https://antonkov.github.io/posts/How-to-build-a-proof-tree/#how-tactics-like-have-fit-into-the-model

- `have H : 0 ≤ (x - y) ^ 2 := by
    positivity` was treated as a single step which
  1. didn't bifurcate the goal
  2.  goal type stayed unchanged.

- whether `have H : 0 ≤ (x - y) ^ 2 
  · positivity` was treated as a sequence of 2 steps.
`have` without `:=` bifurcated the goal into `0 ≤ (x - y) ^ 2` and the one with `H` in the context.
But because of bifurcation the heuristic didn't apply.

**Comment by antonkov** [comment]
![example](https://antonkov.github.io/assets/img/haveprooftree_example.png)

**Comment by lakesare** [comment]
Modern Lean errors out with the second-example code unfortunately: [https://live.lean-lang.org](https://live.lean-lang.org/#code=import%20Mathlib.Tactic.Positivity%0Aimport%20Mathlib.Tactic.Linarith%0Aexample%20%7Bx%20y%20%3A%20%E2%84%9A%7D%20%3A%202%20*%20x%20*%20y%20%E2%89%A4%20x%20%5E%202%20%2B%20y%20%5E%202%20%3A%3D%20by%0A%20%20have%20H%20%3A%200%20%E2%89%A4%20(x%20-%20y)%20%5E%202%20%0A%20%20%C2%B7%20positivity%0A%20%20linarith) (`unexpected token '·'; expected '|'`).

I will close this issue for now - if you have a way to recreate a similar issue in modern Lean, please reopen.

**Comment by hrmacbeth** [comment]
I believe it's just that in current mathlib one needs another import:
```
import Mathlib.Tactic.Have
```

**Comment by lakesare** [comment]
From what I see, Paperproof renders this appropriately. These are mathematically similar statements, but they are structurally different - and Paperproof accurately reflects that.

When we go through this proof:

```lean
example {x y : ℚ} : 2 * x * y ≤ x ^ 2 + y ^ 2 := by
  have H : 0 ≤ (x - y) ^ 2
  · positivity
  linarith
```

letter by letter, we do have two goals at some point:

<img width="246" alt="image" src="https://github.com/user-attachments/assets/51f69059-7060-4aee-830c-231dee50e1f5">


And this is reflected in Paperproof:
<img width="356" alt="image" src="https://github.com/user-attachments/assets/d9b670ef-6582-4cc1-abae-68a691a0c439">


The second proof is rendered differently than the first one because it offers different functionality. For example, in the second case, the user can type `pick_goal 2` (and that's obvious from the way paperproof renders it), while in the first case they cannot do that. Also, the visualisation of the second proof makes it obvious what the role of `·` exactly is.
I think it makes sense to follow exactly what's going on in the proof internally. I think rather than aiming for the prettiest possible visualisation of a proof, we want to aim for high trust that "paperproof will display the proof as it really is internally".

Open to objections though.

**Comment by hrmacbeth** [comment]
That seems reasonable. Thanks for thinking about it.
